crates/wasm-shrink/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+//! Shrink a Wasm file while maintaining a property of interest (such as
+//! triggering a compiler bug).
+//!
+//! See the [`WasmShrink`][WasmShrink] type for details.
+
+#![deny(missing_docs)]
+
+use anyhow::{Context, Result};
+use is_executable::IsExecutable;
+use rand::{rngs::SmallRng, Rng, SeedableRng};
+use std::path::{Path, PathBuf};
+use wasm_mutate::WasmMutate;
+
+#[rustfmt::skip]
+static EMPTY_WASM: &'static [u8] = &[
+    // Magic.
+    0x00, b'a', b's', b'm',
+    // Version.
+    0x01, 0x00, 0x00, 0x00,
+];
+
+/// Shrink a Wasm file while maintaining a property of interest (such as
+/// triggering a compiler bug).
+///
+/// # Example
+///
+/// Given that
+///
+/// * `crasher.wasm` is a Wasm file that triggers a crash in your Wasm compiler
+///   when you try to compile it, and
+///
+/// * `compile.sh` is a script that exits with code zero if its first
+///   command-line argument triggers a crash in your Wasm compiler, and exits
+///   with a non-zero code otherwise.
+///
+/// then you can shrink `crasher.wasm` into a smaller Wasm file that still
+/// triggers the crash and save it at `shrunken.wasm` with the following code
+/// snippet:
+///
+/// ```
+/// use wasm_shrink::WasmShrink;
+///
+/// # fn foo() -> anyhow::Result<()> {
+/// WasmShrink::new("compile.sh", "crasher.wasm")
+///     .output("shrunken.wasm")
+///     .run()?;
+/// # Ok(()) }
+/// ```
+#[cfg_attr(feature = "structopt", derive(structopt::StructOpt))]
+#[cfg_attr(
+    feature = "structopt",
+    structopt(
+        about = r###"
+Shrink a Wasm file while maintaining a property of interest (such as triggering
+a compiler bug).
+"###,
+        long_about = r###"
+Shrink a Wasm file while maintaining a property of interest (such as triggering
+a compiler bug).
+
+## Example
+
+Given that
+
+* `crasher.wasm` is a Wasm file that triggers a crash in your Wasm compiler when
+  you try to compile it, and
+
+* `compile.sh` is a script that exits with code zero if its first command-line
+  argument triggers a crash in your Wasm compiler, and exits with a non-zero
+  code otherwise.
+
+then you can shrink `crasher.wasm` into a smaller Wasm file that still triggers
+the crash and save it at `shrunken.wasm` with the following command:
+
+    $ wasm-shrink compile.sh crasher.wasm -o shrunken.wasm
+"###
+    )
+)]
+pub struct WasmShrink {
+    /// The output file path to write the shrunken Wasm file to.
+    ///
+    /// By default, a file path based on the input will be generated.
+    #[cfg_attr(feature = "structopt", structopt(short, long))]
+    output: Option<PathBuf>,
+
+    /// The number of shrink attempts to try before considering a Wasm module as
+    /// small as it will ever get.
+    #[cfg_attr(feature = "structopt", structopt(short, long, default_value = "1000"))]
+    attempts: u32,
+
+    /// The RNG seed for choosing which size-reducing mutation to attempt next.
+    #[cfg_attr(feature = "structopt", structopt(short, long, default_value = "42"))]
```

Comment:
I'm not familiar with general shrinking techniques, but would it make sense for the default value to be a random number instead?

Reply:
I don't think it matters either way, and this way we can get deterministic shrinking by default, which is relatively rare and nice to have.

Reply:
I think determinism will mostly be nice for testing.

But either way, we have to provide a seed to `wasm-mutate` so we might as well make it customizable. If we want to add parallelism in the future, we can give up the determinism. But also parallelism doesn't help that much for these kinds of brute force, greedy search algorithms in my experience.